src/core/namespace.c

```diff
@@ -665,40 +674,45 @@ static int make_read_only(BindMount *m, char **blacklist) {
 
 static int chase_all_symlinks(const char *root_directory, BindMount *m, unsigned *n) {
         BindMount *f, *t;
-        int r;
+        int r = 0;
 
         assert(m);
-        assert(n);
+
+        /* Nothing to do */
+        if (!*n)
+                return r;
 
         /* Since mount() will always follow symlinks and we need to take the different root directory into account we
          * chase the symlinks on our own first. This call wil do so for all entries and remove all entries where we
          * can't resolve the path, and which have been marked for such removal. */
 
         for (f = m, t = m; f < m + *n; f++) {
                 _cleanup_free_ char *chased = NULL;
-
                 r = chase_symlinks(f->path, root_directory, &chased);
                 if (r == -ENOENT && f->ignore) {
                         /* Doesn't exist? Then remove it! */
                         f->path = mfree(f->path);
+                        r = 0;
                         continue;
                 }
-                if (r < 0)
-                        return log_debug_errno(r, "Failed to chase symlinks for %s: %m", f->path);
+                if (r < 0) {
+                        log_debug_errno(r, "Failed to chase symlinks for %s: %m", f->path);
+                        break;
+                }
```

Comment:
I don't think this fix is correct either. The issue with double free came from the fact that if we broke iteration in the middle, same path would be pointed at from two places: `*t` and `*f`, and then when doing the cleanup, we'd attempt to free it from both.

 `n_mounts` is used *in the caller* to check how many mounts should be freed. So to ensure that we can correctly free all allocations on error, we cannot store anything shorter than the original value of `*n`, unless we have iterated over the whole array or otherwise ensure that we have no duplicate or stale pointers. Just iterating until the end seems like an easier option, and this is not code that requires optimization. 

So:
1. please add a temporary `k` variable.
2. do `if (k < 0 && r >= 0) r = k` (remember earliest error)
3. continue until the end and return `r`, also setting `*n`.
This way the cleanup loop can do it's job.